### PR TITLE
Rework MultibootImage to use ReaderAt instead of filename

### PR DESCRIPTION
**Pull request by JulienVdG**
The goal is to unify the interface with LinuxImage and hopefully allow to PXE boot multiboot kernel

### CODE REVIEW

**Comment by JulienVdG** [comment]
I wonder if I should move the Gzip filter out, like in pkg/uio and have it used by the loaders

File: pkg/boot/multiboot/reader.go [hunk 1]
```diff
@@ -60,3 +67,14 @@ func readFile(name string) ([]byte, error) {
 
 	return ioutil.ReadAll(f)
 }
+
+// TODO: could be tryCompressionFilters, grub support gz,xz and lzop
+// TODO: do we want to keep the filter inside multiboot? This could be the responsibility of the caller...
+func tryGzipFilter(r io.ReaderAt) io.ReaderAt {
+	b, err := readGzip(uio.Reader(r))
+	if err == nil {
+		kernel := kernelReader{buf: b}
```
Comment (hugelgupf) [inline pkg/boot/multiboot/reader.go]: why is kernelReader a thing? could it just be bytes.Buffer?

File: pkg/bootconfig/bootconfig.go [hunk 2]
```diff
@@ -87,12 +88,31 @@ func (bc *BootConfig) Boot() error {
...
+
 		// check multiboot header
-		if err := multiboot.Probe(bc.Multiboot); err != nil {
+		if err := multiboot.Probe(mbkernel); err != nil {
 			log.Printf("Error parsing multiboot header: %v", err)
 			return err
 		}
-		if err := multiboot.Load(true, bc.Multiboot, bc.MultibootArgs, bc.Modules, nil); err != nil {
+		modules := make([]multiboot.Module, len(bc.Modules))
+		for i, cmd := range bc.Modules {
```
Comment (hugelgupf) [inline pkg/bootconfig/bootconfig.go]: i feel like this whole opening files act is in a few places... should pkg/boot/multiboot have some utilities for dealing with it?

File: pkg/boot/multiboot/reader.go (hunk 1 above)
Comment (JulienVdG) [reply pkg/boot/multiboot/reader.go]: I kept the existing code, but you're right I should simplify it, I'll try bytes.Buffer, that sounds good (and already implement the inMemReaderAt interface) !

File: pkg/bootconfig/bootconfig.go (hunk 2 above)
Comment (JulienVdG) [reply pkg/bootconfig/bootconfig.go]: yeah propably, I was thinking 'diskboot' ... but mutliboot probably makes sense, like 
```
func openModules(cmds []string) ([]Modules, error)
```

Comment (JulienVdG) [reply pkg/bootconfig/bootconfig.go]: or even a `LoadFromFiles` with the previous prototype of `Load`

Comment (hugelgupf) [reply pkg/bootconfig/bootconfig.go]: I think the former would be nice, since it's usable with boot.MultibootImage

Comment (JulienVdG) [reply pkg/bootconfig/bootconfig.go]: done

File: pkg/boot/multiboot/reader.go (hunk 1 above)
Comment (JulienVdG) [reply pkg/boot/multiboot/reader.go]: Except bytes.Buffer is not a ReaderAt, bytes.Reader is one (but not a inMemReaderAt)...
I replaced the optimization with using io.WriterTo interface when possible.

**Comment by JulienVdG** [comment]
This should be ready now (and CI did run this time)

File: pkg/boot/kexec/memory_linux.go [hunk 3]
```diff
@@ -365,13 +365,18 @@ func (m *Memory) LoadElfSegments(r io.ReaderAt) error {
...
-		if err != nil {
-			return err
-		}
-		if n < len(d) {
-			return fmt.Errorf("not all data of the segment was read")
+		var d []byte
+		if p.Filesz == 0 {
+			d = make([]byte, p.Memsz)
+		} else {
+			d = make([]byte, p.Filesz)
```
Comment (hugelgupf) [inline pkg/boot/kexec/memory_linux.go]: Are you sure about this? When memsz > filesz, I thought it's supposed to fill the rest with zeros, which is a behavior not retained in this PR

File: pkg/boot/multiboot/module.go [hunk 4]
```diff
@@ -102,22 +102,29 @@ func alignUp(buf *bytes.Buffer) error {
...
 	// place start of each module to a beginning of a page.
 	if err := alignUp(buf); err != nil {
 		return err
 	}
 
 	m.Start = uint32(buf.Len())
-	if _, err := buf.Write(b); err != nil {
-		return err
+
+	if wtr, ok := r.(io.WriterTo); ok {
```
Comment (hugelgupf) [inline pkg/boot/multiboot/module.go]: Just use io.Copy(buf, r)? io.Copy does all this under the hood afaik.

File: pkg/boot/kexec/memory_linux.go (hunk 3 above)
Comment (hugelgupf) [reply pkg/boot/kexec/memory_linux.go]: oh, nevermind. i got it. had to read the code below. notice how the segment is made with Paddr and Memsz, not Filesz.

You don't need to have `d` be a slice of 0s -- kexec pads segments that have bufsz < memsz with 0s. So a kexec_segment as below fills the purpose, and `d` can be `nil`, which should be converted to `buf: 0, bufsz: 0`.

{
buf: 0
bufsz: 0
mem: 0x1000
memsz: 0x1000
}

the kernel knows how to interpret that in kexec - fill the area with 0s.

Did you encounter a specific bug with this?

Comment (hugelgupf) [reply pkg/boot/kexec/memory_linux.go]: Perhaps it'd be a good idea to quote the man-page in the code for future readers:

```
bufsz bytes are copied from the source buffer to the target kernel
          buffer.  If bufsz is less than memsz, then the excess bytes in the
          kernel buffer are zeroed out.
```

http://man7.org/linux/man-pages/man2/kexec_load.2.html

Comment (JulienVdG) [reply pkg/boot/kexec/memory_linux.go]: Thanks for the detailed analysis, I didn't know the details of kexec load.

I had a bug with my build of the multiboot test kernel (build with buildroot so a different gcc and libc). The resulting elf is triggering it with a section that was misinterpreted by the parser.
I initially got the error `error loading ELF segments: bad offset 8192` but as the size is 0, the offset is irrelevant.

File: pkg/boot/multiboot/module.go (hunk 4 above)
Comment (JulienVdG) [reply pkg/boot/multiboot/module.go]: Thanks, I often miss some basic functions from the standard library (probably a bias from my C background)

**Pull request merged**